### PR TITLE
Multi-GPU FSDP Support

### PR DESCRIPTION
**Pull request by humaira-rf**
### Summary
- Integrated multi-GPU training via Fully Sharded Data Parallelism (FSDP), enabling large models to be distributed across multiple GPUs.

### Key changes:
- Added FSDP support for sharding model parameters, gradients, and optimizer states across GPUs
- Extended RFModelConfig and experiment.run_fit APIs to accept a num_gpus argument
- Updated scheduler logic to allocate runs across multiple GPUs

### Testing:

- Single-GPU (regression): Ran all original lite notebooks - no regressions
- Multi-GPU FSDP:
  
  1. Ran lite, standard, and large FSDP notebooks on downsampled data
  2. Validated across multiple FSDP configurations (full shard, CPU offload, gradient checkpointing), multiple config leaves.
  3. Tested all ICOps - resume, clone, delete, clone (warm-start)
  4. Verified training plots and metric logging render correctly

### CODE REVIEW

File: rapidfireai/fit/backend/controller.py [hunk 1]
```diff
@@ -217,7 +294,12 @@ def _process_interactive_control(
                     status=RunStatus.STOPPED,
                     ended_by=RunEndedBy.INTERACTIVE_CONTROL,
                 )
-                self.db.set_ic_ops_task_status(run_state["task_id"], TaskStatus.COMPLETED)
+                self.db.set_ic_ops_task_status(
+                    run_state["task_id"], TaskStatus.COMPLETED
+                )
+                self.db.set_ic_ops_task_status(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: Aren't lines 300-302 is just repeating 297-299?

File: rapidfireai/fit/backend/controller.py [hunk 2]
```diff
@@ -234,7 +316,12 @@ def _process_interactive_control(
                     status=RunStatus.DELETED,
                     ended_by=RunEndedBy.INTERACTIVE_CONTROL,
                 )
-                self.db.set_ic_ops_task_status(run_state["task_id"], TaskStatus.COMPLETED)
+                self.db.set_ic_ops_task_status(
+                    run_state["task_id"], TaskStatus.COMPLETED
+                )
+                self.db.set_ic_ops_task_status(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 3 LOC repeated again

File: rapidfireai/fit/backend/controller.py [hunk 3]
```diff
@@ -243,11 +330,21 @@ def _process_interactive_control(
                     status=RunStatus.ONGOING,
                     ended_by="",
                 )
-                self.db.set_ic_ops_task_status(run_state["task_id"], TaskStatus.COMPLETED)
+                self.db.set_ic_ops_task_status(
+                    run_state["task_id"], TaskStatus.COMPLETED
+                )
+                self.db.set_ic_ops_task_status(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 3 LOC repeated again

File: rapidfireai/fit/backend/controller.py [hunk 4]
```diff
@@ -243,11 +330,21 @@ def _process_interactive_control(
...
+                    run_state["task_id"], TaskStatus.COMPLETED
+                )
                 self.ic_logger.info(f"Resuming run {run_id} by Interactive Control")
             elif run_state["status"] == RunStatus.COMPLETED:
                 # process completed tasks
-                self.logger.warning(f"Run {run_id} is already completed. Skipping Interactive Control task.")
+                self.logger.warning(
+                    f"Run {run_id} is already completed. Skipping Interactive Control task."
+                )
+                self.logger.warning(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 3 LOC repeated again

File: rapidfireai/fit/backend/controller.py [hunk 5]
```diff
@@ -263,7 +360,12 @@ def _process_interactive_control(
             # add additional_kwargs to config_leaf if it exists in the parent run
             parent_run_details = self.db.get_run(parent_run_id)
             if "additional_kwargs" in parent_run_details["config_leaf"]:
-                config_leaf["additional_kwargs"] = parent_run_details["config_leaf"]["additional_kwargs"]
+                config_leaf["additional_kwargs"] = parent_run_details["config_leaf"][
+                    "additional_kwargs"
+                ]
+                config_leaf["additional_kwargs"] = parent_run_details["config_leaf"][
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 3 LOC repeated again

File: rapidfireai/fit/backend/controller.py [hunk 6]
```diff
@@ -321,8 +433,18 @@ def _process_interactive_control(
...
                 self.db.set_ic_ops_task_status(task["task_id"], TaskStatus.FAILED)
-                self.ic_logger.error(f"Error creating model for run {parent_run_id}: {e}")
-                raise ControllerException(f"Error creating model for run {parent_run_id}: {e}") from e
+                self.ic_logger.error(
+                    f"Error creating model for run {parent_run_id}: {e}"
+                )
+                raise ControllerException(
+                    f"Error creating model for run {parent_run_id}: {e}"
+                ) from e
+                self.ic_logger.error(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 6 LOC repeated again

File: rapidfireai/fit/backend/controller.py [hunk 7]
```diff
@@ -376,21 +504,35 @@ def _process_interm_ic_ops_states(
                     ControllerTask.IC_STOP,
                 ]:
                     # ignore RESUME/STOP tasks for completed runs
-                    self.ic_logger.warning(f"Ignoring RESUME/STOP task for run {run_id} as it is already completed")
+                    self.ic_logger.warning(
+                        f"Ignoring RESUME/STOP task for run {run_id} as it is already completed"
+                    )
+                    self.ic_logger.warning(
```
Comment (arun-rfai) [inline rapidfireai/fit/backend/controller.py]: 3 LOC repeated again

File: tutorial_notebooks/fine-tuning/rf-tutorial-sft-chatqa-fsdp-large.ipynb [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        "        training_args=RFSFTConfig(\n",
+        "            learning_rate=1e-4,  \n",
+        "            lr_scheduler_type=\"linear\",\n",
+        "            per_device_train_batch_size=1,\n",
+        "            per_device_eval_batch_size=1,\n",
+        "            num_train_epochs=1,\n",
+        "            gradient_accumulation_steps=4,   \n",
+        "            logging_steps=1,\n",
+        "            eval_strategy=\"steps\",\n",
+        "            eval_steps=1,\n",
```
Comment (chethan-rfai) [inline tutorial_notebooks/fine-tuning/rf-tutorial-sft-chatqa-fsdp-large.ipynb]: Can the eval_steps updated to large number (>8) to speed up the training time
